x-pack/plugins/task_manager/server/lib/throughput_manager.ts

```diff
@@ -0,0 +1,173 @@
+    ).pipe(
+      // When tag is "flush", reset the error counter
+      // Otherwise increment the error counter
+      mergeScan(
+        ({ count }, next) =>
+          next === FLUSH_MARKER
+            ? of(emitErrorCount(count), resetErrorCount())
+            : of(incementErrorCount(count)),
+        emitErrorCount(0)
+      ),
+      filter(isEmitEvent),
+      map(({ count }) => count)
+    );
+  }
+
+  public get isStarted() {
+    return this.throughputCheckSubcription && !this.throughputCheckSubcription.closed;
+  }
+
+  public start() {
+    if (!this.isStarted) {
+      this.throughputCheckSubcription = this.throughputCheck$.subscribe((errorCount) => {
+        if (errorCount > 0) {
+          this.reduceThroughput(errorCount);
+          return;
+        }
+        this.increaseThroughput();
+      });
+    }
+  }
+
+  public stop() {
+    this.throughputCheckSubcription?.unsubscribe();
+    // Reset observable values to original values
+    this.updateConfiguration(this.maxWorkers.startingValue, this.pollInterval.startingValue);
+  }
+  private reduceThroughput(errorCount: number) {
+    const newMaxWorkers = Math.max(
+      Math.floor(this.maxWorkers.currentValue * MAX_WORKERS_DECREASE_PERCENTAGE),
+      1
+    );
+    const newPollInterval = Math.ceil(
+      this.pollInterval.currentValue * POLL_INTERVAL_INCREASE_PERCENTAGE
+    );
+    this.logger.info(
+      `Throughput reduced after seeing ${errorCount} error(s): maxWorkers: ${this.maxWorkers.currentValue}->${newMaxWorkers}, pollInterval: ${this.pollInterval.currentValue}->${newPollInterval}`
+    );
+    this.updateConfiguration(newMaxWorkers, newPollInterval);
+  }
+
+  private increaseThroughput() {
+    const newMaxWorkers = Math.min(
+      this.maxWorkers.startingValue,
+      Math.ceil(this.maxWorkers.currentValue * MAX_WORKERS_INCREASE_PERCENTAGE + 1)
+    );
+    const newPollInterval = Math.max(
+      this.pollInterval.startingValue,
+      Math.floor(this.pollInterval.currentValue * POLL_INTERVAL_DECREASE_PERCENTAGE - 1)
+    );
+    if (
+      newMaxWorkers !== this.maxWorkers.currentValue ||
+      newPollInterval !== this.pollInterval.currentValue
+    ) {
+      this.logger.info(
+        `Throughput increasing after seeing no errors: maxWorkers: ${this.maxWorkers.currentValue}->${newMaxWorkers}, pollInterval: ${this.pollInterval.currentValue}->${newPollInterval}`
+      );
+    }
+    this.updateConfiguration(newMaxWorkers, newPollInterval);
+  }
+
+  private updateConfiguration(newMaxWorkers: number, newPollInterval: number) {
+    if (this.maxWorkers.currentValue !== newMaxWorkers) {
+      this.maxWorkers.currentValue = newMaxWorkers;
+      this.maxWorkers.observable$.next(newMaxWorkers);
+    }
+    if (this.pollInterval.currentValue !== newPollInterval) {
+      this.pollInterval.currentValue = newPollInterval;
+      this.pollInterval.observable$.next(newPollInterval);
+    }
```

Comment:
It would be nice if, instead of pushing into a Subject, the throughputManager returns two streams, one for workers and one for interval, and we subscribed to those where we need these values, rather than passing around the Subjects and having their values pushed in from multiple places. 🤔 

This is not a must at all, but I think, overall, it would make the composition easier to follow.